### ISSUE TITLE
Bug: Hitting backspace deletes two characters on Android 

### ISSUE THREAD

**Issue by fredthebuilder** [facebook/lexical#4941]
Lexical version: 0.11.3 & 0.12.0 (also on https://playground.lexical.dev/)

## Steps To Reproduce

1. Open https://playground.lexical.dev/ on Chrome on Android
2. Write a word
3. Put the cursor somewhere in the middle
4. hit backspace - two characters get deleted

## The current behavior

https://github.com/facebook/lexical/assets/143589809/3660f9cc-19eb-4764-8b03-3c9a2549e089

(results in HelloWorld -> Hellorld instead of HelloWorld -> HellWorld)

Two characters get deleted

## The expected behavior

Only the character before the cursor gets deleted

**Comment by moy2010** [comment]
This is a known issue: https://github.com/facebook/lexical/issues/4340

**Comment by kirandash** [comment]
> In this thread there's a solution which is not perfect but it kinda works https://github.com/facebook/lexical/issues/3896
> 
> 
> 
> I created a patch using `patch-package` where I commented out the `else` block as described in one of the comments. However it produces couple of really weird side effects (which I believe are connected):
> 
> - When you enter just one character and try to delete it - it simply doesn't do it on Android, cursor moves to the beginning of the string leaving you with that letter being intact
> 
> - When you try do select all and delete - again doesn't do anything, except of moving cursor to the beginning of the string
> 
> - When you try to remove mention nodes and have a custom handling for this scenario - it doesn't do so correctly, hard to explain but maybe it's just my case
> 
> 
> 
> I'm currently working on finding a workaround, but so far no progress. I'm thinking of just adding a "trash icon" near the editor on all Android devices to help users clear up data in case they can't achieve it with backspace

Hi @kazimirchuk did you manage to find a fix for this? The current fix from lexical produces these weird side effects as you mentioned which is a bad UX at the moment

**Comment by ivailop7** [comment]
Yes, it's merged, but we haven't published a lexical version since, so the fix is not out yet. That said, this is more of a short-term patch, until the actual Chromium mobile bug gets sorted out.

**Comment by kirandash** [comment]
Thanks for your response @kazimirchuk and @ivailop7 🙏🏻 Much appreciated 🙂

Unfortunately this issue is a blocker for me now since the editor on my project is for end users especially on mobile device. I am curious how the facebook team managed to patch this for the post editor.

**Comment by kirandash** [comment]
I updated my code to the following to solve the last character not being deleted. Tested on chrome, safari, android, iPhone and desktop and seems to be fine and I don't see any side effects. Hope it helps someone. Any feedback is much appreciated :)

https://github.com/facebook/lexical/assets/13310363/c4313b45-798b-4502-b20b-0fa41b3316c6

```
if ($isRangeSelection(selection)) {
    // Used for handling backspace in Android.
    // Fix for bug: backspace in Android deleting two characters at once https://github.com/facebook/lexical/issues/3896 
    // PR: https://github.com/facebook/lexical/pull/5077/files/db8a6cd3ae9d38884b12b44a15289dc980784e67
    const IS_ANDROID = CAN_USE_DOM && /Android/.test(navigator.userAgent);
    if (IS_ANDROID) {
      $setCompositionKey(selection.anchor.key);
    }
    if (
      isPossiblyAndroidKeyPress(event.timeStamp) &&
      editor.isComposing() &&
      selection.anchor.key === selection.focus.key
    ) {
      $setCompositionKey(null);
      lastKeyDownTimeStamp = 0;
      // Fixes an Android bug where selection flickers when backspacing
      setTimeout(() => {
        updateEditor(editor, () => {
          $setCompositionKey(null);
        });
      }, ANDROID_COMPOSITION_LATENCY);
      if ($isRangeSelection(selection)) {
        const anchorNode = selection.anchor.getNode();
        anchorNode.markDirty();
        selection.format = anchorNode.getFormat();
        selection.style = anchorNode.getStyle();
      }
      // Fix for bug: backspace in Android not deleting first character https://github.com/facebook/lexical/pull/5077#issuecomment-1771409941
      const selectedText = selection.anchor.getNode().getTextContent()
      if (selectedText.length <= 1) {
        event.preventDefault();
        dispatchCommand(editor, DELETE_CHARACTER_COMMAND, true);
      }
    } else {
      event.preventDefault();
      dispatchCommand(editor, DELETE_CHARACTER_COMMAND, true);
    }
    return;
  }
}
```

cc @kazimirchuk @ivailop7

**Comment by kirandash** [comment]
> @kirandash I just tried out your fix and it works! 🎉 Thanks a lot for sharing! 🙇‍♂️ It's a major improvement. Maybe it can also be merged into main? I don't see a big difference in approach between this and manual set of composition key 🤷
> 
> Worth noting that there's one more issue with removals (not related to your change at all):
> 
> * Select All
> * Hit "Delete"
> 
> Expected: All text is removed What happened: None of the text is removed and cursor jumps to the beginning of the text
> 
> But all in all I think deleting all is a very rare case and can be achieved via char by char removal

Glad that it helped :) And thanks a lot for sharing the issue on select all and delete. I will also have a look.

**Comment by bnol** [comment]
> > @kirandash I just tried out your fix and it works! 🎉 Thanks a lot for sharing! 🙇‍♂️ It's a major improvement. Maybe it can also be merged into main? I don't see a big difference in approach between this and manual set of composition key 🤷
> > Worth noting that there's one more issue with removals (not related to your change at all):
> > 
> > * Select All
> > * Hit "Delete"
> > 
> > Expected: All text is removed What happened: None of the text is removed and cursor jumps to the beginning of the text
> > But all in all I think deleting all is a very rare case and can be achieved via char by char removal
> 
> Glad that it helped :) And thanks a lot for sharing the issue on select all and delete. I will also have a look. And regarding if it should be merged to main? It seemed more like a patch than a fix hence I didn't create a PR. But you are right, it's similar approach to the manual set of composition key. So I guess I will create a PR soon

@kirandash thanks for the fix. 
Is there any idea that I can manually apply your fix while waiting for your code to be merged and released?

**Comment by kirandash** [comment]
> > > @kirandash I just tried out your fix and it works! 🎉 Thanks a lot for sharing! 🙇‍♂️ It's a major improvement. Maybe it can also be merged into main? I don't see a big difference in approach between this and manual set of composition key 🤷
> > > Worth noting that there's one more issue with removals (not related to your change at all):
> > > 
> > > * Select All
> > > * Hit "Delete"
> > > 
> > > Expected: All text is removed What happened: None of the text is removed and cursor jumps to the beginning of the text
> > > But all in all I think deleting all is a very rare case and can be achieved via char by char removal
> > 
> > 
> > Glad that it helped :) And thanks a lot for sharing the issue on select all and delete. I will also have a look. And regarding if it should be merged to main? It seemed more like a patch than a fix hence I didn't create a PR. But you are right, it's similar approach to the manual set of composition key. So I guess I will create a PR soon
> 
> @kirandash thanks for the fix. Is there any idea that I can manually apply your fix while waiting for your code to be merged and released?

@bnol I have now created a PR here: https://github.com/facebook/lexical/pull/5282 that you can follow for the latest updates. In the meantime, you can use a patch ex: `patch-package` for npm or `pnpm patch` etc to patch Lexical.dev.js and Lexical.prod.js files on your project.

How to generate the Lexical.dev.js and Lexical.prod.js file with the fix?
Go to LexicalEvent.ts file: add the fix code and run `npm run build-release`. Once the build is ready, use those files from dist folder on your project.

Hope that helps.

**Comment by amanharwara** [comment]
> Worth noting that there's one more issue with removals (not related to your change at all):
> 
>     * Select All
> 
>     * Hit "Delete"
> 
> 
> Expected: All text is removed What happened: None of the text is removed and cursor jumps to the beginning of the text

@kazimirchuk I noticed this only happens if the selection's `anchor.key` and `focus.key` are the same, if they're not it leads to #5259. From what I've been able to understand, this is because the fix in #5077 works by setting the composition key so that it goes to the [if-branch](https://github.com/facebook/lexical/blob/main/packages/lexical/src/LexicalEvents.ts#L533-L535) where nothing is deleted. One of the conditions there is `anchor.key === focus.key`, which is why if you type a paragraph without any other node in it and then select all + delete, nothing gets deleted because the anchor key and focus key are the same.

**Comment by unsidez** [comment]
Hello!
Any news regarding this?
I've been facing this issue in my project where we have a webview in an app and when user's try to delete some characters, they delete the previous character and the next one.
I was wondering about solutions but didn't found anything.

**Comment by unsidez** [comment]
Hey
Just wanted to say that i've updated the lib to the latest version and I no longer see the issue!
Thanks!
